ambry-store/src/main/java/com.github.ambry.store/StatsManager.java

```diff
@@ -0,0 +1,206 @@
+    }
+  }
+
+  /**
+   * Fetch and aggregate stats from a given {@link Store}
+   * @param aggregatedSnapshot the {@link StatsSnapshot} to hold the aggregated result
+   * @param partitionId specifies the {@link Store} to be fetched from
+   * @return true if stats were successfully fetched from the given {@link PartitionId}, false otherwise
+   */
+  boolean collectAndAggregate(StatsSnapshot aggregatedSnapshot, PartitionId partitionId) {
+    long fetchAndAggregatePerStoreStartTimeMs = time.milliseconds();
+    Store store = storageManager.getStore(partitionId);
+    if (store == null) {
+      return false;
+    } else {
+      StatsSnapshot statsSnapshot = null;
+      try {
+        statsSnapshot = store.getStoreStats().getStatsSnapshot();
+      } catch (StoreException e) {
+        logger.error("Store exception thrown when getting stats for partitionId: {}", partitionId, e);
+        return false;
+      }
+      aggregate(aggregatedSnapshot, statsSnapshot);
+      metrics.fetchAndAggregateTimePerStore.update(time.milliseconds() - fetchAndAggregatePerStoreStartTimeMs);
+      return true;
+    }
+  }
+
+  /**
+   * Performs recursive aggregation of two {@link StatsSnapshot} and stores the result in the first one.
+   * @param baseSnapshot one of the addends and where the result will be
+   * @param newSnapshot the other addend to be added into the first {@link StatsSnapshot}
+   */
+  private void aggregate(StatsSnapshot baseSnapshot, StatsSnapshot newSnapshot) {
+    baseSnapshot.setValue(baseSnapshot.getValue() + newSnapshot.getValue());
+    if (baseSnapshot.getSubtree() == null) {
+      baseSnapshot.setSubtree(newSnapshot.getSubtree());
+    } else if (newSnapshot.getSubtree() != null) {
+      for (Map.Entry<String, StatsSnapshot> entry : newSnapshot.getSubtree().entrySet()) {
+        if (!baseSnapshot.getSubtree().containsKey(entry.getKey())) {
+          baseSnapshot.getSubtree().put(entry.getKey(), new StatsSnapshot(0L, null));
+        }
+        aggregate(baseSnapshot.getSubtree().get(entry.getKey()), entry.getValue());
+      }
+    }
+  }
+
+  /**
+   * Runnable class that collects, aggregate and publish stats via methods in StatsManager.
+   */
+  private class StatsAggregator implements Runnable {
+    private boolean cancelled = false;
+
+    @Override
+    public void run() {
+      long totalFetchAndAggregateStartTimeMs = time.milliseconds();
+      StatsSnapshot aggregatedSnapshot = new StatsSnapshot(0L, null);
+      List<String> unreachableStores = new ArrayList<>();
+      Iterator<PartitionId> iterator = initialPartitionIds.iterator();
+      while (!cancelled && iterator.hasNext()) {
+        PartitionId partitionId = iterator.next();
+        if (!collectAndAggregate(aggregatedSnapshot, partitionId)) {
+          unreachableStores.add(partitionId.toString());
+        }
+      }
+      if (!cancelled) {
+        metrics.totalFetchAndAggregateTime.update(time.milliseconds() - totalFetchAndAggregateStartTimeMs);
```

Comment:
I decided to put it inside so the total fetch and aggregate time won't be skewed when aggregator thread is cancelled (will report significantly shorter time).